diff --git a/equi_diffpo/__init__.py b/equi_diffpo/__init__.py
new file mode 100644
index 0000000..e69de29
diff --git a/equi_diffpo/common/robomimic_util.py b/equi_diffpo/common/robomimic_util.py
index 030cc2e..e52e2dc 100644
--- a/equi_diffpo/common/robomimic_util.py
+++ b/equi_diffpo/common/robomimic_util.py
@@ -38,7 +38,7 @@ class RobomimicAbsoluteActionConverter:
             render_offscreen=False,
             use_image_obs=False, 
         )
-        assert not abs_env.env.robots[0].controller.use_delta
+        assert not abs_env.env.robots[0].controller.use_relative
 
         self.env = env
         self.abs_env = abs_env
diff --git a/equi_diffpo/config/train_equi_diffusion_unet_abs.yaml b/equi_diffpo/config/train_equi_diffusion_unet_abs.yaml
index 1827c77..2ca3f9a 100644
--- a/equi_diffpo/config/train_equi_diffusion_unet_abs.yaml
+++ b/equi_diffpo/config/train_equi_diffusion_unet_abs.yaml
@@ -17,7 +17,7 @@ n_latency_steps: 0
 dataset_obs_steps: ${n_obs_steps}
 past_action_visible: False
 dataset: equi_diffpo.dataset.robomimic_replay_image_sym_dataset.RobomimicReplayImageSymDataset
-dataset_path: data/robomimic/datasets/${task_name}/${task_name}_abs.hdf5
+dataset_path: ../../datasets/${task_name}/${task_name}_abs.hdf5
 
 policy:
   _target_: equi_diffpo.policy.diffusion_equi_unet_cnn_enc_policy.DiffusionEquiUNetCNNEncPolicy
diff --git a/equi_diffpo/env_runner/robomimic_image_runner.py b/equi_diffpo/env_runner/robomimic_image_runner.py
index 70dcd15..4139289 100644
--- a/equi_diffpo/env_runner/robomimic_image_runner.py
+++ b/equi_diffpo/env_runner/robomimic_image_runner.py
@@ -22,7 +22,7 @@ from equi_diffpo.env.robomimic.robomimic_image_wrapper import RobomimicImageWrap
 import robomimic.utils.file_utils as FileUtils
 import robomimic.utils.env_utils as EnvUtils
 import robomimic.utils.obs_utils as ObsUtils
-
+import mimicgen # noqa: F401
 
 def create_env(env_meta, shape_meta, enable_render=True):
     modality_mapping = collections.defaultdict(list)
diff --git a/equi_diffpo/model/equi/equi_obs_encoder_no_eih.py b/equi_diffpo/model/equi/equi_obs_encoder_no_eih.py
new file mode 100644
index 0000000..cef299d
--- /dev/null
+++ b/equi_diffpo/model/equi/equi_obs_encoder_no_eih.py
@@ -0,0 +1,109 @@
+import torch
+from torchvision import models as vision_models
+from escnn import gspaces, nn
+from escnn.group import CyclicGroup
+from einops import rearrange
+from robomimic.models.base_nets import SpatialSoftmax
+from equi_diffpo.model.common.module_attr_mixin import ModuleAttrMixin
+import equi_diffpo.model.vision.crop_randomizer as dmvc
+from equi_diffpo.model.equi.equi_encoder import EquivariantResEncoder76Cyclic, EquivariantVoxelEncoder58Cyclic, EquivariantVoxelEncoder64Cyclic
+from equi_diffpo.model.vision.voxel_crop_randomizer import VoxelCropRandomizer
+from equi_diffpo.model.common.rotation_transformer import RotationTransformer
+
+class Identity(torch.nn.Module):
+    def __init__(self, *args, **kwargs):
+        super(Identity, self).__init__()
+
+    def forward(self, x):
+        return x
+    
+class InHandEncoder(torch.nn.Module):
+    def __init__(self, out_size):
+        super().__init__()
+        net = vision_models.resnet18(norm_layer=Identity)
+        self.resnet = torch.nn.Sequential(*(list(net.children())[:-2]))
+        self.spatial_softmax = SpatialSoftmax([512, 3, 3], num_kp=out_size//2)
+
+    def forward(self, ih):
+        batch_size = ih.shape[0]
+        return self.spatial_softmax(self.resnet(ih)).reshape(batch_size, -1)
+
+
+class EquivariantObsEnc(ModuleAttrMixin):
+    def __init__(
+        self,
+        obs_shape=(3, 84, 84),
+        crop_shape=(76, 76),
+        n_hidden=128,
+        N=8,
+        initialize=True,
+    ):
+        super().__init__()
+        obs_channel = obs_shape[0]
+        self.n_hidden = n_hidden
+        self.N = N
+        self.group = gspaces.no_base_space(CyclicGroup(self.N))
+        self.token_type = nn.FieldType(self.group, self.n_hidden * [self.group.regular_repr])
+        self.enc_obs = EquivariantResEncoder76Cyclic(obs_channel, self.n_hidden, initialize)
+        self.enc_out = nn.Linear(
+            nn.FieldType(
+                self.group,
+                n_hidden * [self.group.regular_repr] # agentview
+                + 4 * [self.group.irrep(1)] # pos, rot
+                + 3 * [self.group.trivial_repr], # gripper (2), z zpos
+            ),
+            self.token_type,
+        )
+        
+        self.quaternion_to_sixd = RotationTransformer('quaternion', 'rotation_6d')
+
+        self.gTgc = torch.Tensor([[1, 0, 0], [0, -1, 0], [0, 0, -1]])
+
+        self.crop_randomizer = dmvc.CropRandomizer(
+            input_shape=obs_shape,
+            crop_height=crop_shape[0],
+            crop_width=crop_shape[1],
+        )
+
+    def get6DRotation(self, quat):
+        # data is in xyzw, but rotation transformer takes wxyz
+        return self.quaternion_to_sixd.forward(quat[:, [3, 0, 1, 2]]) 
+        
+    def forward(self, nobs):
+        obs = nobs["agentview_image"]
+        ee_pos = nobs["robot0_eef_pos"]
+        ee_quat = nobs["robot0_eef_quat"]
+        ee_q = nobs["robot0_gripper_qpos"]
+        # B, T, C, H, W
+        batch_size = obs.shape[0]
+        t = obs.shape[1]
+        obs = rearrange(obs, "b t c h w -> (b t) c h w")
+        ee_pos = rearrange(ee_pos, "b t d -> (b t) d")
+        ee_quat = rearrange(ee_quat, "b t d -> (b t) d")
+        ee_q = rearrange(ee_q, "b t d -> (b t) d")
+        obs = self.crop_randomizer(obs)
+        ee_rot = self.get6DRotation(ee_quat)
+
+        enc_out = self.enc_obs(obs).tensor.reshape(batch_size * t, -1)  # b d
+        pos_xy = ee_pos[:, 0:2]
+        pos_z = ee_pos[:, 2:3]
+        features = torch.cat(
+            [
+                enc_out,
+                pos_xy,
+                # ee_rot is the first two rows of the rotation matrix (i.e., the rotation 6D repr.)
+                # each column vector in the first two rows of the rotation 6d forms a rho1 vector
+                ee_rot[:, 0:1],
+                ee_rot[:, 3:4],
+                ee_rot[:, 1:2],
+                ee_rot[:, 4:5],
+                ee_rot[:, 2:3],
+                ee_rot[:, 5:6],
+                pos_z,
+                ee_q,
+            ],
+            dim=1
+        )
+        features = nn.GeometricTensor(features, self.enc_out.in_type)
+        out = self.enc_out(features).tensor
+        return rearrange(out, "(b t) d -> b t d", b=batch_size)
\ No newline at end of file
diff --git a/equi_diffpo/policy/diffusion_equi_unet_cnn_enc_policy.py b/equi_diffpo/policy/diffusion_equi_unet_cnn_enc_policy.py
index c0d176a..9f2bc5c 100644
--- a/equi_diffpo/policy/diffusion_equi_unet_cnn_enc_policy.py
+++ b/equi_diffpo/policy/diffusion_equi_unet_cnn_enc_policy.py
@@ -13,7 +13,7 @@ try:
         raise ImportError("CropRandomizer is not in robomimic.models.base_nets")
 except ImportError:
     import robomimic.models.obs_core as rmbn
-from equi_diffpo.model.equi.equi_obs_encoder import EquivariantObsEnc
+from equi_diffpo.model.equi.equi_obs_encoder_no_eih import EquivariantObsEnc
 from equi_diffpo.model.equi.equi_conditional_unet1d import EquiDiffusionUNet
 from equi_diffpo.model.vision.rot_randomizer import RotRandomizer
 
diff --git a/equi_diffpo/scripts/robomimic_dataset_conversion.py b/equi_diffpo/scripts/robomimic_dataset_conversion.py
index dd4adb1..206aba7 100644
--- a/equi_diffpo/scripts/robomimic_dataset_conversion.py
+++ b/equi_diffpo/scripts/robomimic_dataset_conversion.py
@@ -16,6 +16,8 @@ from tqdm import tqdm
 import collections
 import pickle
 from equi_diffpo.common.robomimic_util import RobomimicAbsoluteActionConverter
+import mimicgen # noqa: F401
+
 
 def worker(x):
     path, idx, do_eval = x
